src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -846,29 +858,44 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                 self.implicit_region_bound,
                 self.param_env,
                 locations,
+                category,
                 &mut borrowck_context.constraints.outlives_constraints,
                 &mut borrowck_context.constraints.type_tests,
                 &mut borrowck_context.all_facts,
             ).convert_all(&data);
         }
     }
 
-    fn sub_types(&mut self, sub: Ty<'tcx>, sup: Ty<'tcx>, locations: Locations) -> Fallible<()> {
+    fn sub_types(
+        &mut self,
+        sub: Ty<'tcx>,
+        sup: Ty<'tcx>,
+        locations: Locations,
+        category: ConstraintCategory,
+    ) -> Fallible<()> {
         relate_tys::sub_types(
             self.infcx,
             sub,
             sup,
             locations,
+            category,
             self.borrowck_context.as_mut().map(|x| &mut **x),
         )
     }
 
-    fn eq_types(&mut self, a: Ty<'tcx>, b: Ty<'tcx>, locations: Locations) -> Fallible<()> {
+    fn eq_types(
+        &mut self,
+        a: Ty<'tcx>,
+        b: Ty<'tcx>,
+        locations: Locations,
+        category: ConstraintCategory,
```

Comment:
So, I did not want to do it this way — but I'm trying to remember *why* and coming up empty. Seems ok.

I'm wondering if it makes sense to keep *both* locations and category as distinct things. Perhaps so: in particular, locations (once we transition to polonius) will be relevant to whether or not this is an error, whereas `category` is not.